---
title: "iRep_PTMETA"
author: "Julian Trachsel"
date: "10/25/2019"
output: html_document
---


# OUTLINE
  1) EXP design
  Turkeys, in-feed antibiotics, continuous treatment throughout
      3 treatments
        control
        subther
        ther
      3 timepoints
        7
        35
        78
      10 birds per treatment per timepoint = 90 birds total
      no repeated measurements, (cecal contents are a terminal sample)
  2) data available
      shotgun metagenomic approach
      cecal contents at necropsy
        hiseq 2x150 PE reads 
          1 library per bird 
          READ STATISTICS
        PacBio RSII reads
          library preps from pooled samples
            1 library prep per timepoint
              equal pool of all treatments
              3 PacBio RSII library preps
              3 SMRT cells?
            READ STATISTICS
  3) Metagenomic assembly with (Meta)Spades
      One large co-assembly, all reads pooled
      Read QC and normalization
        Share QC slurm script
      Metaspades used, can align long reads to assembly graph
        Share metaspades script
        ~50 hours on mem node
      ASSEMBLY STATS
  4) Genome binning using metabat
      map reads to co-assembly
        SHARE READ MAPPING SCRIPT
      convert to indexed bams
      feed indexed bams and co-assembly to metabat
        SHARE METABAT SCRIPT
  5) Filter bins based on CheckM requirements
      Only high quality bins can be used
      90% complete
      less than 2% contamination
      not too fragmented
  6) Calculate iRep
      inputs:
        Mapped sam files generated by mapping all reads back to co-assembly
        folder of filtered genomes (bins)
        SHARE iREP script
  7) R for plots and stats
      SHARE R SCRIPT
      
        